README.md
# Ruby OpenMPT (ffi-openmpt)
## Robert Haines

### Synopsys

A ruby interface to `libopenmpt` - part of [OpenMPT][mpt-home].

See the [libopenmpt homepage][lib-home] for more information.

[![Gem Version](https://badge.fury.io/rb/ffi-openmpt.svg)](https://badge.fury.io/rb/ffi-openmpt)
[![Build Status](https://travis-ci.org/hainesr/ffi-openmpt.svg?branch=master)](https://travis-ci.org/hainesr/ffi-openmpt)
[![Maintainability](https://api.codeclimate.com/v1/badges/919bd8b421798dbd2719/maintainability)](https://codeclimate.com/github/hainesr/ffi-openmpt/maintainability)
[![Coverage Status](https://coveralls.io/repos/github/hainesr/ffi-openmpt/badge.svg?branch=master)](https://coveralls.io/github/hainesr/ffi-openmpt)

### Prerequisits

#### libopenmpt

You must have `libopenmpt` installed. On Ubuntu this is done with:

```shell
$ sudo apt install libopenmpt0
```

You do not need the `libopenmpt` development libraries to be installed. If this library fails to find `libopenmpt` on your platform it might be due to it being named something slightly different. Please [raise an issue][issues] to let me know.

Instructions for installing `libopenmpt` from source are available on the [libopenmpt homepage][lib-home].

#### Ruby FFI

This library uses [Ruby FFI][ruby-ffi] (foreign function interface) to wrap the C API for `libopenmpt`. It will be installed automatically if you install this package via [Ruby Gems][rubygems], otherwise:

```shell
$ gem install ffi
```

will do the trick.

### Installation

Add `ffi-openmpt` to your `.gemspec` or `Gemfile` as appropriate, or install directly from [Ruby Gems][rubygems]:

```shell
$ gem install ffi-openmpt
```

### Usage

The library wraps the C `libopenmpt` API directly: methods have the same name and signature as their C counterparts. A more friendly ruby-like interface is also available, which hides the FFI details as much as possible.

Not all `libopenmpt` methods are wrapped yet, but enough functionality is supplied to load a module, interogate it and render it to a PCM stream.

#### A note on strings returned by `libopenmpt`

`libopenmpt` manages the memory of any strings it returns. This means that you must free up such memory explicitly after you have finished with them when using the C API. Such strings are returned to ruby as [`FFI::Pointer`][ffi-pointer] objects, so the string value can be copied to a ruby string as follows:

```ruby
begin
  ptr = FFI::OpenMPT::API.openmpt_get_string('url')
  str = ptr.read_string
ensure
  FFI::OpenMPT::API.openmpt_free_string(ptr)
end

puts str
```

The ruby interface handles all this for you:

```ruby
str = FFI::OpenMPT::String.get(:url)
puts str
```

### Example scripts

Scripts in the `examples` directory show how to use both the C and ruby APIs. You will need to make sure that `ffi-openmpt` is on your `RUBYLIB` path, or run the examples with `bundle exec`.

#### `mod-info` and `mod-info-api`

Display information about a mod file, for example:

```shell
$ ./mod-info lastsun.mod

Ruby OpenMPT (ffi-openmpt) file interrogator.
---------------------------------------------

Filename...: lastsun.mod
Size.......: 106k
Type.......: mod
Format.....: Generic Amiga / PC MOD file
Tracker....: Master Soundtracker 1.0
Title......: the last sun
Duration...: 3:56.4
Subsongs...: 1
Channels...: 4
Orders.....: 35
Patterns...: 20
Intruments.: 0
Samples....: 15
```

Both `mod-info` and `mod-info-api` output exactly the same data. `mod-info` uses the ruby interface, and `mod-info-api` uses the mapped `libopenmpt` C API directly.

#### `mod-2-raw`

Render a mod to a raw stereo PCM file. Use the `--float` switch to generate 32 bit float data, or omit for 16bit int data. For example:

```shell
$ ./mod-2-raw --float lastsun.mod

Ruby OpenMPT (ffi-openmpt) Mod to Raw PCM Converter.
----------------------------------------------------

Filename...: lastsun.mod
Size.......: 106k
Type.......: mod
Output type: float.raw
Output size: 88687k
```

The raw output format is simply a blob of `float` or `short` data. The left and right stereo channels are interleaved. The sample rate used is 48,000Hz.

### Library versions

Until this library reaches version 1.0.0 the API may be subject to breaking changes. When version 1.0.0 is released, then the principles of [semantic versioning][semver] will be applied.

### Licence

The Ruby OpenMPT code (ffi-openmpt) is released under the BSD licence.

Copyright (c) 2018, 2019, Robert Haines
All rights reserved.

See LICENCE for more details.

### Acknowledgements

For testing purposes, this library uses and distributes ['The Last Sun'][lastsun] by Frederic Hahn. It is believed to be in the Public Domain, but if this is not the case please [raise an issue][issues] to let me know.

[ffi-pointer]: https://www.rubydoc.info/github/ffi/ffi/FFI/Pointer
[issues]: https://github.com/hainesr/ffi-openmpt/issues
[lastsun]: https://modarchive.org/module.php?47521
[lib-home]: https://lib.openmpt.org
[mpt-home]: https://openmpt.org/
[ruby-ffi]: https://rubygems.org/gems/ffi
[rubygems]: https://rubygems.org
[semver]: https://semver.org/


CHANGES.md
# Changes log for the Ruby OpenMPT library (ffi-openmpt)

## Version 0.5.0

* Nail down rubocop version and update todo list.
* Add rubocop-performance to the setup.
* Wrap some libopenmpt playing info functions.
* Add the informational calls to the ruby interface.
* Wrap the speed and tempo libopenmpt functions.
* Add the speed and tempo calls to the ruby interface.
* Wrap the current playing channels libopenmpt function.
* Add current playing channels to the ruby interface.
* Wrap the VU meter libopenmpt functions.
* Add VU meter calls to the ruby interface.

## Version 0.4.1

* Use an explicit begin block in Module#get_names.

## Version 0.4.0

* Update the minimum ruby version to 2.3.
* Update to use bundler 2.0.1.
* Update the string example in README.md.
* Simplify the API string code example.
* Add sample_name to the ruby interface.
* Align info method names with the C++ API.
* Ensure memory is always freed in String::get.
* Ensure memory is always freed in Module methods.
* Ensure memory is always freed in OpenMPT methods.
* Update the memory handling code example in the README.
* Use `warn` not `puts` for error messages in mod-2-raw.
* Fix wording of an error message in mod-2-raw.
* Add Module#sample_names to the ruby interface.
* Add a note in Module::new about MemoryPointer.
* Remove Module#sample_name. It's not in the C++ API.
* Generalize getting name lists from a mod.
* Add Module#instrument_names to the ruby interface.
* Add Module#pattern_names to the ruby interface.
* Add Module#order_names to the ruby interface.
* Add Module#channel_names to the ruby interface.
* Add Module#subsong_names to the ruby interface.

## Version 0.3.0

* Wrap the libopenmpt module read mono functions.
* Add mono read (short/float) calls to the ruby interface.
* Move the string get method to its own namespace.
* Allow getting informational strings directly.
* Wrap Module#metadata_keys.
* Directly query module metadata keys.
* Update test version of libopenmpt to 0.3.13.
* Parameterize the test libopenmpt version.
* Update Travis config to use latest rubies.
* Turn on `fast_finish` in the Travis config.
* Switch to use latest dist (Xenial) in Travis.
* Test multiple versions of the libopenmpt library.
* Wrap libopenmpt module repeat functions.
* Add repeat_count to the ruby interface.
* Wrap libopenmpt module positional calls.
* Add positional calls to the ruby interface.

## Version 0.2.0

* Add a Gem version badge to the README.md.
* Wrap the libopenmpt module probing functions.
* Enforce comparisons, not predicates, for older rubies.
* Wrap libopenmpt informational function: get subsongs.
* Wrap libopenmpt informational function: get channels.
* Wrap libopenmpt informational function: get orders.
* Wrap libopenmpt informational function: get patterns.
* Wrap libopenmpt informational function: get instruments.
* Wrap libopenmpt informational function: get samples.
* Wrap libopenmpt informational function: subsong names.
* Wrap libopenmpt informational function: channel names.
* Wrap libopenmpt informational function: order names.
* Wrap libopenmpt informational function: pattern names.
* Wrap libopenmpt informational function: instrument names.
* Wrap libopenmpt informational function: sample names.
* Wrap libopenmpt render parameter functions.
* Add an example script to read mod information.
* Document the mod-info example in the README.
* Use File.binread to load module data in the example.
* Rename the example script to show that it uses the API.
* Start ruby-like interface to a mod: Module.
* Add the metadata method to the ruby interface.
* Add duration to the ruby interface.
* Add the simple informational calls to the ruby interface.
* Replicate the 'mod-info' example with the ruby interface.
* Add notion of a closed mod to the ruby interface.
* Catch a closed mod and return in the ruby interface.
* Allow access to the metadata via ruby interface methods.
* Use the new metadata method mappings in mod-info.
* Update README with details of the second example script.
* Add the get_string library call to the ruby interface.
* Add the extensions library call to the ruby interface.
* Add the supported_extension? library call to the ruby interface.
* Move the module probe function to the OpenMPT namespace.
* Get render param 'gain' in the ruby interface.
* Set render param 'gain' in the ruby interface.
* Get render param 'stereo_separation' in the ruby interface.
* Set render param 'stereo_separation' in the ruby interface.
* Get render param 'interpolation_filter' in the ruby interface.
* Set render param 'interpolation_filter' in the ruby interface.
* Get render param 'volume_ramping' in the ruby interface.
* Set render param 'volume_ramping' in the ruby interface.
* Add transient_error? call to the ruby interface.
* Add error_string call to the ruby interface.
* Change supported_extensions to return list of Symbols.
* Add sample_rate instance variable to Module.
* Add missing tests for a closed module.
* Add stereo read (short) calls to the ruby interface.
* Add stereo read (float) calls to the ruby interface.
* Add a 'mod-2-raw' ruby interface example script.
* Add a note to the README about finding libopenmpt.
* Update the README with information about the ruby interface.

## Version 0.1.0

* Add Code of Conduct.
* Add Travis configuration.
* Add a Travis badge to the README.
* Set up Coveralls integration.
* Add a Coveralls badge to the README.
* Add a CodeClimate badge to the README.
* Add rubocop configuration files.
* Add rubocop tasks to the Rakefile.
* Many rubocop fixes.
* Add ffi as a dependency.
* Wrap the libopenmpt version methods.
* Wrap the libopenmpt logging callbacks.
* Add tests for the API module.
* Test the logging API functions.
* Wrap the libopenmpt error callbacks.
* Wrap libopenmpt module create/destroy functions.
* Add a helper method to test against mod files.
* Wrap the libopenmpt module duration function.
* Wrap the libopenmpt module read stereo functions.

test/test_helper.rb
# frozen_string_literal: true

# Copyright (c) 2018, 2019, Robert Haines.
#
# Licensed under the BSD License. See LICENCE for details.

require 'coveralls'
Coveralls.wear!

$LOAD_PATH.unshift File.expand_path('../lib', __dir__)
require 'ffi/openmpt'

require 'minitest/autorun'
require 'tempfile'

FIXTURES_DIR = ::File.expand_path('fixtures', __dir__)

# Utility function to ensure that we can capture stderr from the
# external C openmpt library.
def read_stderr
  orig_stderr = STDERR.dup

  Tempfile.open('ffi_openmpt_stderr') do |err|
    STDERR.reopen(err)
    yield
    err.rewind
    err.read
  end
ensure
  STDERR.reopen(orig_stderr)
end

# Test mods
MOD_LAST_SUN = ::File.join(FIXTURES_DIR, 'lastsun.mod')

# Raw data
RAW_LAST_SUN_INT16 = ::File.join(FIXTURES_DIR, 'lastsun.int16.raw')
RAW_LAST_SUN_FLOAT = ::File.join(FIXTURES_DIR, 'lastsun.float.raw')
RAW_LAST_SUN_MONO_INT16 = ::File.join(FIXTURES_DIR, 'lastsun.mono.int16.raw')
RAW_LAST_SUN_MONO_FLOAT = ::File.join(FIXTURES_DIR, 'lastsun.mono.float.raw')

# Data within the test mod
SAMPLE_NAMES_LAST_SUN = [
  'music composed by',
  'fred in march 1989.',
  '',
  'if you need a tune',
  'for a game please',
  'write to',
  '  frederic hahn',
  '  17 rue du haut-barr',
  '  67550 vendenheim',
  '  france.',
  'or call 88-69-40-22',
  '',
  'feel free to rip this',
  'music but please only',
  'use doc 2.0 replayer'
].freeze

METADATA_KEYS_LAST_SUN = [
  [:type, 'mod'],
  [:type_long, 'Generic Amiga / PC MOD file'],
  [:tracker, 'Master Soundtracker 1.0'],
  [:title, 'the last sun'],
  [:message_raw, '']
].freeze


test/openmpt_test.rb
# frozen_string_literal: true

# Copyright (c) 2018, 2019, Robert Haines.
#
# Licensed under the BSD License. See LICENCE for details.

require 'test_helper'

class FFI::OpenMPTTest < Minitest::Test

  def test_the_library_version_number
    version = ::FFI::OpenMPT.library_version
    refute_nil version
    assert_kind_of Array, version
    assert_equal 3, version.length
  end

  def test_the_core_version_number
    version = ::FFI::OpenMPT.core_version
    refute_nil version
    assert_kind_of Array, version
    assert_equal 4, version.length
  end

  def test_supported_extensions
    exts = ::FFI::OpenMPT.supported_extensions
    assert_kind_of Array, exts
    refute_empty exts
    exts.each { |ext| assert_kind_of Symbol, ext }
  end

  def test_extension_supported?
    [:mod, :med, :stm, :xm, :it].each do |ext|
      assert ::FFI::OpenMPT.extension_supported?(ext)
    end

    [:aaa, :bbb, :ccc].each do |ext|
      refute ::FFI::OpenMPT.extension_supported?(ext)
    end
  end

  def test_transient_error
    assert ::FFI::OpenMPT.transient_error?(
      ::FFI::OpenMPT::API::OPENMPT_ERROR_OUT_OF_MEMORY
    )

    refute ::FFI::OpenMPT.transient_error?(
      ::FFI::OpenMPT::API::OPENMPT_ERROR_ARGUMENT_NULL_POINTER
    )
  end

  def test_error_string
    [
      [
        ::FFI::OpenMPT::API::OPENMPT_ERROR_OUT_OF_MEMORY,
        'out of memory'
      ],
      [
        ::FFI::OpenMPT::API::OPENMPT_ERROR_ARGUMENT_NULL_POINTER,
        'unknown error'
      ]
    ].each do |error, message|
      assert_equal ::FFI::OpenMPT.error_string(error), message
    end
  end

  def test_probe_good_file
    assert ::FFI::OpenMPT.probe_file(MOD_LAST_SUN)
  end

  def test_probe_bad_file
    refute ::FFI::OpenMPT.probe_file(RAW_LAST_SUN_INT16)
  end
end


lib/ffi/openmpt/module.rb
# frozen_string_literal: true

# Copyright (c) 2018, 2019, Robert Haines.
#
# Licensed under the BSD License. See LICENCE for details.

module FFI
  module OpenMPT
    class Module

      include API

      attr_reader :sample_rate

      def initialize(filename, sample_rate = 48_000)
        @closed = false
        @mod = read_mod(filename)
        @sample_rate = sample_rate

        # Allocate a reusable single int buffer.
        # This is for use by the 'get_render_params'-type calls.
        # FFI::MemoryPointer is garbage collected automatically.
        @int_value = ::FFI::MemoryPointer.new(:int, 1)
      end

      def self.open(filename)
        m = new(filename)

        if block_given?
          begin
            yield m
          ensure
            m.close
          end
        end

        m
      end

      def sample_rate=(rate)
        @sample_rate = rate if (8_000..192_000).cover?(rate)
      end

      def duration
        return if closed?
        openmpt_module_get_duration_seconds(@mod)
      end

      def num_subsongs
        return if closed?
        openmpt_module_get_num_subsongs(@mod)
      end

      def subsong_names
        get_names(num_subsongs, :openmpt_module_get_subsong_name)
      end

      def num_channels
        return if closed?
        openmpt_module_get_num_channels(@mod)
      end

      def channel_names
        get_names(num_channels, :openmpt_module_get_channel_name)
      end

      def num_orders
        return if closed?
        openmpt_module_get_num_orders(@mod)
      end

      def order_names
        get_names(num_orders, :openmpt_module_get_order_name)
      end

      def num_patterns
        return if closed?
        openmpt_module_get_num_patterns(@mod)
      end

      def pattern_names
        get_names(num_patterns, :openmpt_module_get_pattern_name)
      end

      def num_instruments
        return if closed?
        openmpt_module_get_num_instruments(@mod)
      end

      def instrument_names
        get_names(num_instruments, :openmpt_module_get_instrument_name)
      end

      def num_samples
        return if closed?
        openmpt_module_get_num_samples(@mod)
      end

      def sample_names
        get_names(num_samples, :openmpt_module_get_sample_name)
      end

      def repeat_count
        openmpt_module_get_repeat_count(@mod)
      end

      def repeat_count=(count)
        openmpt_module_set_repeat_count(@mod, count)
      end

      def position
        return if closed?
        openmpt_module_get_position_seconds(@mod)
      end

      def position=(param)
        if param.is_a?(Array)
          openmpt_module_set_position_order_row(@mod, param[0], param[1])
        else
          openmpt_module_set_position_seconds(@mod, param)
        end
      end

      def current_speed
        openmpt_module_get_current_speed(@mod)
      end

      def current_tempo
        openmpt_module_get_current_tempo(@mod)
      end

      def current_order
        openmpt_module_get_current_order(@mod)
      end

      def current_pattern
        openmpt_module_get_current_pattern(@mod)
      end

      def current_row
        openmpt_module_get_current_row(@mod)
      end

      def current_playing_channels
        openmpt_module_get_current_playing_channels(@mod)
      end

      def current_channel_vu_mono(channel)
        openmpt_module_get_current_channel_vu_mono(@mod, channel)
      end

      def current_channel_vu_stereo(channel)
        [
          openmpt_module_get_current_channel_vu_left(@mod, channel),
          openmpt_module_get_current_channel_vu_right(@mod, channel)
        ]
      end

      def metadata_keys
        ptr = openmpt_module_get_metadata_keys(@mod)
        ptr.read_string.split(';').map(&:to_sym)
      ensure
        openmpt_free_string(ptr)
      end

      def metadata(key)
        return if closed? || !metadata_keys.include?(key)

        ptr = openmpt_module_get_metadata(@mod, key.to_s)
        ptr.read_string
      ensure
        openmpt_free_string(ptr)
      end

      def gain
        success = openmpt_module_get_render_param(
          @mod, OPENMPT_MODULE_RENDER_MASTERGAIN_MILLIBEL, @int_value
        )

        success == 1 ? @int_value.read_int : nil
      end

      def gain=(value)
        openmpt_module_set_render_param(
          @mod, OPENMPT_MODULE_RENDER_MASTERGAIN_MILLIBEL, value
        )
      end

      def stereo_separation
        success = openmpt_module_get_render_param(
          @mod, OPENMPT_MODULE_RENDER_STEREOSEPARATION_PERCENT, @int_value
        )

        success == 1 ? @int_value.read_int : nil
      end

      def stereo_separation=(value)
        openmpt_module_set_render_param(
          @mod, OPENMPT_MODULE_RENDER_STEREOSEPARATION_PERCENT, value
        )
      end

      def interpolation_filter
        success = openmpt_module_get_render_param(
          @mod, OPENMPT_MODULE_RENDER_INTERPOLATIONFILTER_LENGTH, @int_value
        )

        success == 1 ? @int_value.read_int : nil
      end

      def interpolation_filter=(value)
        openmpt_module_set_render_param(
          @mod, OPENMPT_MODULE_RENDER_INTERPOLATIONFILTER_LENGTH, value
        )
      end

      def volume_ramping
        success = openmpt_module_get_render_param(
          @mod, OPENMPT_MODULE_RENDER_VOLUMERAMPING_STRENGTH, @int_value
        )

        success == 1 ? @int_value.read_int : nil
      end

      def volume_ramping=(value)
        openmpt_module_set_render_param(
          @mod, OPENMPT_MODULE_RENDER_VOLUMERAMPING_STRENGTH, value
        )
      end

      def read_mono(frames, buffer)
        openmpt_module_read_mono(@mod, @sample_rate, frames, buffer)
      end

      def read_float_mono(frames, buffer)
        openmpt_module_read_float_mono(@mod, @sample_rate, frames, buffer)
      end

      def read_stereo(frames, left, right)
        openmpt_module_read_stereo(@mod, @sample_rate, frames, left, right)
      end

      def read_interleaved_stereo(frames, buffer)
        openmpt_module_read_interleaved_stereo(
          @mod, @sample_rate, frames, buffer
        )
      end

      def read_float_stereo(frames, left, right)
        openmpt_module_read_float_stereo(
          @mod, @sample_rate, frames, left, right
        )
      end

      def read_interleaved_float_stereo(frames, buffer)
        openmpt_module_read_interleaved_float_stereo(
          @mod, @sample_rate, frames, buffer
        )
      end

      def close
        return if closed?
        @closed = true
        openmpt_module_destroy(@mod)
      end

      def closed?
        @closed
      end

      def method_missing(name, *args)
        respond_to?(name) ? metadata(name) : super
      end

      def respond_to_missing?(name, *all)
        metadata_keys.include?(name) || super
      end

      private

      def read_mod(filename)
        data = ::File.binread(filename)
        mod = openmpt_module_create_from_memory2(
          data,
          data.bytesize,
          LogSilent, nil, ErrorIgnore, nil, nil, nil, nil
        )

        @closed = (mod.address == 0)
        mod
      end

      def get_names(num, get)
        (0...num).reduce([]) do |acc, i|
          begin
            ptr = send(get, @mod, i)
            acc << ptr.read_string
          ensure
            openmpt_free_string(ptr)
          end
        end
      end
    end
  end
end


lib/ffi/openmpt/string.rb
# frozen_string_literal: true

# Copyright (c) 2018, 2019, Robert Haines.
#
# Licensed under the BSD License. See LICENCE for details.

module FFI
  module OpenMPT
    module String
      KEYS = [
        :library_version,
        :library_version_major,
        :library_version_minor,
        :library_version_patch,
        :library_version_prerel,
        :library_version_is_release,
        :library_features,
        :core_version,
        :source_url,
        :source_date,
        :source_revision,
        :source_is_modified,
        :source_has_mixed_revisions,
        :source_is_package,
        :build,
        :build_compiler,
        :credits,
        :contact,
        :license,
        :url,
        :support_forum_url,
        :bugtracker_url
      ].freeze

      def self.get(key)
        ptr = API.openmpt_get_string(key.to_s)
        ptr.read_string
      ensure
        API.openmpt_free_string(ptr)
      end

      def self.method_missing(name, *args)
        respond_to?(name) ? get(name) : super
      end

      def self.respond_to_missing?(name, *all)
        KEYS.include?(name) || super
      end
    end
  end
end


lib/ffi/openmpt/openmpt.rb
# frozen_string_literal: true

# Copyright (c) 2018, 2019, Robert Haines.
#
# Licensed under the BSD License. See LICENCE for details.

module FFI
  module OpenMPT
    def self.library_version
      maj, min, pat = [API.openmpt_get_library_version].pack('L>').unpack('CCS')
      [maj, min, pat >> 8]
    end

    def self.core_version
      [API.openmpt_get_core_version].pack('L>').unpack('CCCC')
    end

    def self.supported_extensions
      ptr = API.openmpt_get_supported_extensions
      ptr.read_string.split(';').map(&:to_sym)
    ensure
      API.openmpt_free_string(ptr)
    end

    def self.extension_supported?(ext)
      API.openmpt_is_extension_supported(ext.to_s) == 1
    end

    def self.transient_error?(error)
      API.openmpt_error_is_transient(error) == 1
    end

    def self.error_string(error)
      ptr = API.openmpt_error_string(error)
      ptr.read_string
    ensure
      API.openmpt_free_string(ptr)
    end

    def self.probe_file(filename)
      probe_size = API.openmpt_probe_file_header_get_recommended_size
      data = ::File.binread(filename, probe_size)
      data_size = ::File.size(filename)
      probe_result = API.openmpt_probe_file_header(
        API::OPENMPT_PROBE_FILE_HEADER_FLAGS_DEFAULT,
        data,
        data.bytesize,
        data_size,
        API::LogSilent, nil, API::ErrorIgnore, nil, nil, nil
      )

      probe_result == API::OPENMPT_PROBE_FILE_HEADER_RESULT_SUCCESS
    end
  end
end


lib/ffi/openmpt/api.rb
# frozen_string_literal: true

# Copyright (c) 2018, 2019, Robert Haines.
#
# Licensed under the BSD License. See LICENCE for details.

module FFI
  module OpenMPT
    module API
      extend FFI::Library

      ffi_lib 'libopenmpt.so.0'

      # Top-level (library) informational API calls
      attach_function :openmpt_get_library_version, [], :uint
      attach_function :openmpt_get_core_version, [], :uint
      attach_function :openmpt_free_string, [:pointer], :void
      attach_function :openmpt_get_string, [:string], :pointer
      attach_function :openmpt_get_supported_extensions, [], :pointer
      attach_function :openmpt_is_extension_supported, [:string], :int

      # Logging API calls
      LogDefault =
        attach_function :openmpt_log_func_default, [:string, :pointer], :void
      LogSilent =
        attach_function :openmpt_log_func_silent, [:string, :pointer], :void

      # Error handling API calls
      OPENMPT_ERROR_OK                     = 0
      OPENMPT_ERROR_BASE                   = 256

      OPENMPT_ERROR_UNKNOWN                = OPENMPT_ERROR_BASE + 1
      OPENMPT_ERROR_EXCEPTION              = OPENMPT_ERROR_BASE + 11
      OPENMPT_ERROR_OUT_OF_MEMORY          = OPENMPT_ERROR_BASE + 21

      OPENMPT_ERROR_RUNTIME                = OPENMPT_ERROR_BASE + 30
      OPENMPT_ERROR_RANGE                  = OPENMPT_ERROR_BASE + 31
      OPENMPT_ERROR_OVERFLOW               = OPENMPT_ERROR_BASE + 32
      OPENMPT_ERROR_UNDERFLOW              = OPENMPT_ERROR_BASE + 33

      OPENMPT_ERROR_LOGIC                  = OPENMPT_ERROR_BASE + 40
      OPENMPT_ERROR_DOMAIN                 = OPENMPT_ERROR_BASE + 41
      OPENMPT_ERROR_LENGTH                 = OPENMPT_ERROR_BASE + 42
      OPENMPT_ERROR_OUT_OF_RANGE           = OPENMPT_ERROR_BASE + 43
      OPENMPT_ERROR_INVALID_ARGUMENT       = OPENMPT_ERROR_BASE + 44

      OPENMPT_ERROR_GENERAL                = OPENMPT_ERROR_BASE + 101
      OPENMPT_ERROR_INVALID_MODULE_POINTER = OPENMPT_ERROR_BASE + 102
      OPENMPT_ERROR_ARGUMENT_NULL_POINTER  = OPENMPT_ERROR_BASE + 103

      attach_function :openmpt_error_is_transient, [:int], :int
      attach_function :openmpt_error_string, [:int], :pointer

      OPENMPT_ERROR_FUNC_RESULT_NONE    = 0
      OPENMPT_ERROR_FUNC_RESULT_LOG     = (1 << 0)
      OPENMPT_ERROR_FUNC_RESULT_STORE   = (1 << 1)
      OPENMPT_ERROR_FUNC_RESULT_DEFAULT =
        (OPENMPT_ERROR_FUNC_RESULT_LOG | OPENMPT_ERROR_FUNC_RESULT_STORE)

      ErrorDefault =
        attach_function :openmpt_error_func_default, [:int, :pointer], :int
      ErrorLog =
        attach_function :openmpt_error_func_log, [:int, :pointer], :int
      ErrorStore =
        attach_function :openmpt_error_func_store, [:int, :pointer], :int
      ErrorIgnore =
        attach_function :openmpt_error_func_ignore, [:int, :pointer], :int

      # Module API calls
      # Creational module calls
      attach_function :openmpt_module_create_from_memory2,
                      [
                        :pointer, :int, :pointer, :pointer, :pointer,
                        :pointer, :pointer, :pointer, :pointer
                      ],
                      :pointer
      attach_function :openmpt_module_destroy, [:pointer], :void

      # Informational module calls
      attach_function :openmpt_module_get_duration_seconds, [:pointer], :double
      attach_function :openmpt_module_get_metadata_keys, [:pointer], :pointer
      attach_function :openmpt_module_get_metadata,
                      [:pointer, :string], :pointer
      attach_function :openmpt_module_get_num_subsongs, [:pointer], :int
      attach_function :openmpt_module_get_subsong_name,
                      [:pointer, :int], :pointer
      attach_function :openmpt_module_get_num_channels, [:pointer], :int
      attach_function :openmpt_module_get_channel_name,
                      [:pointer, :int], :pointer
      attach_function :openmpt_module_get_num_orders, [:pointer], :int
      attach_function :openmpt_module_get_order_name,
                      [:pointer, :int], :pointer
      attach_function :openmpt_module_get_num_patterns, [:pointer], :int
      attach_function :openmpt_module_get_pattern_name,
                      [:pointer, :int], :pointer
      attach_function :openmpt_module_get_num_instruments, [:pointer], :int
      attach_function :openmpt_module_get_instrument_name,
                      [:pointer, :int], :pointer
      attach_function :openmpt_module_get_num_samples, [:pointer], :int
      attach_function :openmpt_module_get_sample_name,
                      [:pointer, :int], :pointer

      # Error module calls
      attach_function :openmpt_module_error_get_last, [:pointer], :int
      attach_function :openmpt_module_error_set_last, [:pointer, :int], :void
      attach_function :openmpt_module_error_get_last_message,
                      [:pointer], :pointer
      attach_function :openmpt_module_error_clear, [:pointer], :void

      # Probe module calls
      OPENMPT_PROBE_FILE_HEADER_FLAGS_NONE          = 0x0
      OPENMPT_PROBE_FILE_HEADER_FLAGS_MODULES       = 0x1
      OPENMPT_PROBE_FILE_HEADER_FLAGS_CONTAINERS    = 0x2
      OPENMPT_PROBE_FILE_HEADER_FLAGS_DEFAULT       =
        (OPENMPT_PROBE_FILE_HEADER_FLAGS_MODULES |
          OPENMPT_PROBE_FILE_HEADER_FLAGS_CONTAINERS)

      OPENMPT_PROBE_FILE_HEADER_RESULT_FAILURE      = 0
      OPENMPT_PROBE_FILE_HEADER_RESULT_SUCCESS      = 1
      OPENMPT_PROBE_FILE_HEADER_RESULT_WANTMOREDATA = -1
      OPENMPT_PROBE_FILE_HEADER_RESULT_ERROR        = -255

      attach_function :openmpt_probe_file_header_get_recommended_size, [], :int
      attach_function :openmpt_probe_file_header,
                      [
                        :uint, :pointer, :int, :uint, :pointer, :pointer,
                        :pointer, :pointer, :pointer, :pointer
                      ],
                      :int

      # Render param module calls
      OPENMPT_MODULE_RENDER_MASTERGAIN_MILLIBEL        = 1
      OPENMPT_MODULE_RENDER_STEREOSEPARATION_PERCENT   = 2
      OPENMPT_MODULE_RENDER_INTERPOLATIONFILTER_LENGTH = 3
      OPENMPT_MODULE_RENDER_VOLUMERAMPING_STRENGTH     = 4

      attach_function :openmpt_module_get_render_param,
                      [:pointer, :int, :pointer], :int
      attach_function :openmpt_module_set_render_param,
                      [:pointer, :int, :int], :int

      # Repeat module calls
      attach_function :openmpt_module_get_repeat_count, [:pointer], :int
      attach_function :openmpt_module_set_repeat_count, [:pointer, :int], :void

      # Positional module calls
      attach_function :openmpt_module_get_position_seconds, [:pointer], :double
      attach_function :openmpt_module_set_position_seconds,
                      [:pointer, :double], :double
      attach_function :openmpt_module_set_position_order_row,
                      [:pointer, :int, :int], :double

      # Real-time playing information calls
      attach_function :openmpt_module_get_current_speed, [:pointer], :int
      attach_function :openmpt_module_get_current_tempo, [:pointer], :int
      attach_function :openmpt_module_get_current_order, [:pointer], :int
      attach_function :openmpt_module_get_current_pattern, [:pointer], :int
      attach_function :openmpt_module_get_current_row, [:pointer], :int
      attach_function :openmpt_module_get_current_playing_channels,
                      [:pointer], :int

      # VU meter calls
      attach_function :openmpt_module_get_current_channel_vu_mono,
                      [:pointer, :int], :float
      attach_function :openmpt_module_get_current_channel_vu_left,
                      [:pointer, :int], :float
      attach_function :openmpt_module_get_current_channel_vu_right,
                      [:pointer, :int], :float

      # Read module calls
      attach_function :openmpt_module_read_mono,
                      [:pointer, :int, :int, :pointer],
                      :int
      attach_function :openmpt_module_read_float_mono,
                      [:pointer, :int, :int, :pointer],
                      :int
      attach_function :openmpt_module_read_stereo,
                      [:pointer, :int, :int, :pointer, :pointer],
                      :int
      attach_function :openmpt_module_read_interleaved_stereo,
                      [:pointer, :int, :int, :pointer],
                      :int
      attach_function :openmpt_module_read_float_stereo,
                      [:pointer, :int, :int, :pointer, :pointer],
                      :int
      attach_function :openmpt_module_read_interleaved_float_stereo,
                      [:pointer, :int, :int, :pointer],
                      :int
    end
  end
end
